### PR TITLE
ranger: add video and pdf preview options

### PR DESCRIPTION
**Pull request by schonfinkel**
###### Motivation for this change

An attempt to add video & pdf previews to ranger via the nix derivation out of the box

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

**Comment by teto** [comment]
I would love to see it installable out of the box. Have you checked the impact on closure size though ? It might be best to provide a module or a wrapper in charge of generating the scope.sh. Would prevent useless rebuilds/caching.

File: pkgs/applications/misc/ranger/default.nix [hunk 1]
```diff
@@ -1,9 +1,13 @@
 { stdenv, lib, fetchFromGitHub, python3Packages, file, less, highlight
-, imagePreviewSupport ? true, w3m ? null}:
+, imagePreviewSupport ? true, w3m ? null
+, pdfPreviewSupport ? true, poppler_utils ? null
```
Comment (toonn) [inline pkgs/applications/misc/ranger/default.nix]: It'd be great if you could add support for `mutool` too, we already do this for text previews. I'll open an issue on this over at ranger/ranger.

File: pkgs/applications/misc/ranger/default.nix [hunk 2]
```diff
@@ -52,6 +58,12 @@ python3Packages.buildPythonApplication rec {
     # give image previews out of the box when building with w3m
     substituteInPlace ranger/config/rc.conf \
       --replace "set preview_images false" "set preview_images true"
+  '' + optionalString pdfPreviewSupport ''
+    # edits scope.sh and enables video previews via ffmpegthumbnailer
```
Comment (toonn) [inline pkgs/applications/misc/ranger/default.nix]: You swapped the comments around it seems.

**Review by toonn — CHANGES_REQUESTED** [review]
One necessary change, the other more a request pending addition to `scope.sh`, all that'd have to change is to allow *either* poppler or mupdf as dependency.

**Review by Ma27 — COMMENTED** [review]
Diff seems fine overall. If the closure size impact is too high, we could also add two packages (e.g. `ranger` and `ranger-full`).

Apart from that I agree with @teto, the optional features should be moved to a wrapper for less rebuilds :)

**Review by teto — CHANGES_REQUESTED** [review]
Makes more sense to go for a module straightaway. Maybe even in home-manager if that's what you use ?

**Comment by mtrsk** [comment]
> Makes more sense to go for a module straightaway. Maybe even in home-manager if that's what you use ?

A pure nix way to config ranger would be awesome, but how would a module for it look like? 

Something like this?
```
programs.ranger = {
  enable = true;
  imagePreviewSupport = true;
  pdfPreviewSupport = true;
  videoPreviewSupport = true;
  extraConfig = {
    commands = <...>;
    rc = <...>;
    rifle = <...>;
    scope = <...>;
  };
};
```

**Comment by teto** [comment]
You don't have to cover all the configurations items from the start. Just enable and maybe `preview= { image .... video ... pdf };` would be enough for a start. Looking at ranger
s manpage, seems like the configuration could be implemented as a plugin (this way no need to patch the default files).

**Comment by mtrsk** [comment]
> You don't have to cover all the configurations items from the start. Just enable and maybe `preview= { image .... video ... pdf };` would be enough for a start. Looking at ranger
> s manpage, seems like the configuration could be implemented as a plugin (this way no need to patch the default files).

I'll take a look at this and other reviews an try to write a wrapper as to improve my nix-fu.

**Comment by Ma27** [comment]
Tbh I'm not sure if that's a good idea: `programs.x` modules are perfectly fine for system-wide configuration but in this case we only use those options to configure the package itself. In this case it might be even harmful as this makes it hard for non-NixOS users (e.g. folks using Nix on Ubuntu) to actually use those features from ranger.

**Comment by teto** [comment]
the changes modify the configuration, we shouldn't have to *reinstall* the package just to modify the config. 
Maybe a wrapper would be better but they are not amazingly user friendly (compared to modules that appear in `man configuration.nix` for instance).

@mtrsk you can look at neovim for a wrapper example.
@ma27 Regarding your remark, maybe it would be best to prepare a home-manager module then ? @mtrsk  do you use home-manager (https://github.com/rycee/home-manager/) ? I so wish both projects could converge.

**Comment by mtrsk** [comment]
> @mtrsk you can look at neovim for a wrapper example.

Thanks, I'll take a look at the neovim wrapper (for learning purposes). 

> the changes modify the configuration, we shouldn't have to _reinstall_ the package just to modify the config.
> Maybe a wrapper would be better but they are not amazingly user friendly (compared to modules that appear in `man configuration.nix` for instance).
> @Ma27 Regarding your remark, maybe it would be best to prepare a home-manager module then ? @mtrsk do you use home-manager (https://github.com/rycee/home-manager/) ? I so wish both projects could converge.

It seems it's best to close this PR, as adding more options seems to bring more harm than benefits (once we take into account user friendliness). Making a home-manager module for ranger makes more sense.

**Comment by teto** [comment]
You can ping me on irc or at home-manager should you have any question, I would really like such a module ;)

**Comment by JohnAZoidberg** [comment]
Has any progress of the home-manager configuration been made?

**Comment by mtrsk** [comment]
@JohnAZoidberg I've never had the time to tackle this

**Pull request closed without merging**